src/txn/txn.c

```diff
@@ -1431,33 +1431,21 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
         WT_ERR(__txn_fixup_hs_update(session, hs_cursor));
 
 prepare_verify:
-    if (EXTRA_DIAGNOSTICS_ENABLED(session, WT_DIAGNOSTIC_PREPARED)) {
-        for (; head_upd != NULL; head_upd = head_upd->next) {
-            /*
-             * Assert if we still have an update from the current transaction that hasn't been
-             * resolved or aborted.
-             */
-            WT_ASSERT_ALWAYS(session,
-              head_upd->txnid == WT_TXN_ABORTED || head_upd->prepare_state == WT_PREPARE_RESOLVED ||
-                head_upd->txnid != txn->id,
+    /*
+     * If we are committing a prepared transaction we can check, to some level, that we resolved the
+     * update chain correctly.
+     */
+    if (EXTRA_DIAGNOSTICS_ENABLED(session, WT_DIAGNOSTIC_PREPARED) && commit) {
+        /*
+         * We need to stop walking as soon as we see an aborted update or an update that doesn't
+         * belong to our transaction as it could be freed concurrently by the update obsolete check
+         * logic.
+         */
+        for (; head_upd != NULL && head_upd->txnid != WT_TXN_ABORTED && head_upd->txnid == txn->id;
```

Comment:
@sueloverso do you think its worth dropping a timing stress here?

Reply:
I think it could be useful unless there is a reason not to. Can the update chain change between setting `head_upd` and the assertion? Then maybe the stress timing point would be helpful. But would the timing stress in here be a huge performance issue. This is a loop walking an update chain so it would be sleeping or yielding all the time. That could make it so that tests take far too long or have an outsized impact.

I'm good with either end result from consideration.

Reply:
I was thinking about this and I don't think a timing stress would make a lot of sense here. Thus far the only issues this loop has found are logic issues with the loop itself. The timing stress would be trying to expose that further but wouldn't likely do much otherwise? (That I know of). I don't want to add an additional stress at this stage for that reason, additionally it would impact the performance of diagnostic builds although that is of less concern.